DAY 4
- Recap
- Cosine similarity function - shows how much similar two vectors are
  1. **cosine = (A.B) / ||A|| x ||B|**|

- Matrix multiply with @ -it is not the individual multiplication, rather the sum  of multiplication of respective rows X cols   
# - What does axis=0 vs axis=1 mean in NumPy? -axis-0 = rows,axis-1 = cols

In [1]:
import numpy as np

In [2]:
# day 4 recap
# cosine = (A.B) / ||A|| x ||B||
def cosine_similarity(v1, v2):
  magnitudeV1 = np.linalg.norm(v1)
  magnitudeV2 = np.linalg.norm(v2)

  cosine = (np.dot(v1,v2)) / (magnitudeV1 * magnitudeV2)
  return cosine

v1 = np.array([1,2,3])
v2 = np.array([4,5,6])
print(cosine_similarity(v1, v2))

#Matrix multiply with @
print(v1 @ v2)

0.9746318461970762
32




# Day 5
- Implement matrix multiplication: first with loops, then with @ — time both
-	Write a normalisation function (row-wise, col-wise) using broadcasting only
-	Solve 2 LeetCode Easy problems using only Python (arrays/strings)


In [3]:
#import numpy as np

# Task 1:
- Implement matrix multiplication: first with loops,
  then with @ — time both

In [4]:
#task 1:
#Implement matrix multiplication: first with loops,
#                   then with @ — time both

a = np.random.randint(10,1000,(100,100))
print(a)

[[315 584 808 ... 533 432 590]
 [922 246 301 ... 121 905 191]
 [704 122 216 ... 258  83  52]
 ...
 [532 661 282 ... 782 972 236]
 [234 928 507 ... 428 783 195]
 [807 949 165 ... 399 400 609]]


In [5]:
b= np.random.randint(20,1000,(100,100))
print(b)

[[580 923 276 ... 781 521 869]
 [889 660 520 ... 963 624 184]
 [172 341 123 ... 760 663 941]
 ...
 [994 458 320 ... 232 200 317]
 [812 662 743 ... 424 346 992]
 [464 208 147 ... 563 271  37]]


In [6]:
result1 = np.zeros((100,100),dtype=int)
print(result1)

[[0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 ...
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]]


In [7]:
a.shape
import time

In [8]:
#matrix multiplication: first with loops

# first loop over rows of a
  #then we need columns of b
   # STORE result[] -> multiply  and sum each pair
    #(So result[0][0] = A row 0 · B column 0:)
start = time.time()
#LOOP 1 OVER rows of a
for i in range(a.shape[0]):
  # loop 2 over columns of b ->b.shape[1] --> b(row,cols)->gives columns
  for j in range(b.shape[1]):
    # loop 3 : shared dimension → range(columns of A) == (rows of B)
    for k in range(a.shape[1]):
      result1[i][j] += a[i][k]*b[k][j]

print(result1,"\n")
print(f"Time taken: {time.time() - start:.4f}s")

start1 = time.time()
print(a@b)
print(f"Time taken: {time.time() - start1 :.4f}s")

[[24314598 22861973 23965933 ... 23730979 26317940 24582986]
 [22914595 21308616 23138917 ... 24555395 26628661 25765535]
 [23327043 21573996 23631550 ... 24205082 25423248 24324367]
 ...
 [26405061 23914665 24738095 ... 24474895 27498502 26060614]
 [26905180 24392408 26950931 ... 27949296 30391197 27677512]
 [27108949 22825308 26540991 ... 26261808 30821549 26362954]] 

Time taken: 1.1626s
[[24314598 22861973 23965933 ... 23730979 26317940 24582986]
 [22914595 21308616 23138917 ... 24555395 26628661 25765535]
 [23327043 21573996 23631550 ... 24205082 25423248 24324367]
 ...
 [26405061 23914665 24738095 ... 24474895 27498502 26060614]
 [26905180 24392408 26950931 ... 27949296 30391197 27677512]
 [27108949 22825308 26540991 ... 26261808 30821549 26362954]]
Time taken: 0.0011s


# Task 2:
-Write a normalisation function (row-wise, col-wise) using broadcasting only

## Normalisation
1. Scaling your data so every feature has the same range. Without it, a feature with values 0-10000 dominates a feature with values 0-1 in ML models

### The formula (z-score normalisation):
- ***normalised = (x - mean) / std***

     1.   height  weight  age
- person1  180    80      25
- person2  165    60      30
- person3  175    70      28

  2. Col-wise (most common in ML) → normalise each FEATURE separately. Height normalised against other heights, weight against weights.
  3. Row-wise → normalise each PERSON's values against each other. Less common.

In [9]:
# Task 2: Write a normalisation function (row-wise, col-wise) using broadcasting only
        #Normalisation
def normalise(a, axis):
    # axis=0 → col-wise (normalise each column)
    # axis=1 → row-wise (normalise each row)

    # step 1: calculate mean along the axis
    mean = np.mean(a, axis,keepdims=True)

    # step 2: calculate std along the axis
    std = np.std(a, axis,keepdims=True)

    # step 3: return (a - mean) / std

    return (a-mean)/std



In [10]:
#test array
w= np.arange(1,13,dtype=float).reshape(4,3)
print(w,"\n")

#calculate the column means
print(np.mean(w,axis=0,keepdims = True))
print(np.mean(w,axis=0),"\n")

#Subtract means from array:
result= (w - np.mean(w, axis=0,keepdims=True))
print(result,"\n")

#divide result by std
norm = result/np.std(w, axis=0, keepdims=True)
print(norm)

[[ 1.  2.  3.]
 [ 4.  5.  6.]
 [ 7.  8.  9.]
 [10. 11. 12.]] 

[[5.5 6.5 7.5]]
[5.5 6.5 7.5] 

[[-4.5 -4.5 -4.5]
 [-1.5 -1.5 -1.5]
 [ 1.5  1.5  1.5]
 [ 4.5  4.5  4.5]] 

[[-1.34164079 -1.34164079 -1.34164079]
 [-0.4472136  -0.4472136  -0.4472136 ]
 [ 0.4472136   0.4472136   0.4472136 ]
 [ 1.34164079  1.34164079  1.34164079]]


In [11]:
#test normalisation function
normalise(w,axis=0)  #columns

array([[-1.34164079, -1.34164079, -1.34164079],
       [-0.4472136 , -0.4472136 , -0.4472136 ],
       [ 0.4472136 ,  0.4472136 ,  0.4472136 ],
       [ 1.34164079,  1.34164079,  1.34164079]])

In [12]:
normalise(w,axis=1) #rows

array([[-1.22474487,  0.        ,  1.22474487],
       [-1.22474487,  0.        ,  1.22474487],
       [-1.22474487,  0.        ,  1.22474487],
       [-1.22474487,  0.        ,  1.22474487]])

In [13]:
print(np.mean(normalise(w, axis=0), axis=0))  # should be ~0
print(np.mean(normalise(w, axis=1), axis=1))

[0. 0. 0.]
[0. 0. 0. 0.]


## Day 5 done
- practice below

In [14]:
print("----------------------------------------------------------------------------------")

----------------------------------------------------------------------------------


In [15]:
a1 = np.arange(1,21).reshape(4,5)
b1 = np.arange(11,31).reshape(5,4)
r = np.zeros((4,4))
print("A \n",a1)
print("B \n",b1)
print("R \n",r)

A 
 [[ 1  2  3  4  5]
 [ 6  7  8  9 10]
 [11 12 13 14 15]
 [16 17 18 19 20]]
B 
 [[11 12 13 14]
 [15 16 17 18]
 [19 20 21 22]
 [23 24 25 26]
 [27 28 29 30]]
R 
 [[0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]]


In [16]:
print(b1.shape[0])
print(b1.shape[1])

5
4


In [17]:
a1[[0],[1]]

array([2])

In [18]:
print(len(a1))

4


In [19]:
print(a1@b1)

[[ 325  340  355  370]
 [ 800  840  880  920]
 [1275 1340 1405 1470]
 [1750 1840 1930 2020]]


In [20]:
#Implement matrix multiplication: first with loops
a1[0][0]

for i in range(a1.shape[0]):
  for j in range(b1.shape[1]):
    for k in range(a1.shape[1]):
      r [i][j]+= a1[i][k] * b1[k][j]

print(r)

[[ 325.  340.  355.  370.]
 [ 800.  840.  880.  920.]
 [1275. 1340. 1405. 1470.]
 [1750. 1840. 1930. 2020.]]


Three nested loops:

- Loop 1 → rows of A
- Loop 2 → columns of B
- Loop 3 → multiply and sum each pair

In [21]:
# loop through rows of A

  # LOOP for columns of B
      ## read a[i][j]
      # read b[j][i]

      #Loop 3 → multiply and sum each pair
        ## result += a[i][j]*b[j][i]
